### PR TITLE
Change `secondaryOptions` argument type of the `Rule` function

### PR DESCRIPTION
**Pull request by Mouvedia**
> Which issue, if any, is this issue related to?

N/A

> Is there anything in the PR that needs further explanation?

While changing the type of `RuleMeta['fixable']`, I stumbled on a typo in `selector-class-pattern/index.mjs` which lead me to a wrongly defined type in `index.d.ts`.

This PR:

- fixes a typo
- changes `secondaryOptions`
- aligns `value-no-vendor-prefix` with the other 2 rules
- blocks #7954

Only 3 rules were using the second parameter—`S`:
- `selector-disallowed-list`
- `selector-class-pattern`
- `value-no-vendor-prefix`

`RuleBase` is internal so it is safe to update.
i.e. `Rule` wouldn't be

### CODE REVIEW

**Comment by ybiquitous** [comment]
Is this PR a refactoring or not? I don't understand the reason to change the `fixable` type.

**Comment by Mouvedia** [comment]
> Is this PR a refactoring or not?

It is not.

> I don't understand the reason to change the fixable type.

So that we can introspect/document rules that have partial fixers.
i.e. rules that only have a fixer for certain options
Right now, `fixable` is not granular enough.
It doesn't have much consequences because `fixable: {}` remains truthy.
i.e. `fixable: { foo: false, bar: false }` is `fixable: false` which I have never seen
tl;dr: the only test that I could add would be one that checks whether at least one property is true when fixable is an object

**Comment by ybiquitous** [comment]
Can you please separate the `fixable` change from other changes? I think we should consider the `fixable` change as a minor update because the `fixable?: true` type is exposed both the types and documents.

**Comment by Mouvedia** [comment]
> Can you please separate the `fixable` change from other changes?

OK

**Comment by Mouvedia** [comment]
> Can you please separate the `fixable` change from other changes?

I suppose you mean 2 PRs?

**Comment by ybiquitous** [comment]
Maybe 2 or 3 PRs. I think this PR consists of three parts: `primaryOption`, `secondaryOptions`, and `fixable`. We may combine `primaryOption` and `secondaryOptions`

Also, since the `Rule` type is documented below, I think a changelog item is necessary:

https://github.com/stylelint/stylelint/blob/27b8e16df61e69f0883ea62d220bb39ffc147c70/docs/developer-guide/plugins.md?plain=1#L37-L38

**Comment by Mouvedia** [comment]
> Maybe 2 or 3 PRs. I think this PR consists of three parts: `primaryOption`, `secondaryOptions`, and `fixable`. We may combine `primaryOption` and `secondaryOptions`.
> 
> Also, since the `Rule` type is documented below, I think a changelog item is necessary:
> 
> https://github.com/stylelint/stylelint/blob/27b8e16df61e69f0883ea62d220bb39ffc147c70/docs/developer-guide/plugins.md?plain=1#L37-L38

There is no changes for `primaryOption` because it always fallback to `any`.
i.e. `P extends Primary ? P : any`
By separating the `fixable` change, it's harder to tell why it matters to have `Primary` because `RuleMeta` is not generic anymore.

I agree that by having a separate `fixable` PR, this PR will be easier to review.
tl;dr: only `secondaryOptions` needs a changeset

File: .changeset/poor-boats-reflect.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"stylelint": minor
+---
+
+Changed: `secondaryOptions` type
```
Comment (Mouvedia) [inline .changeset/poor-boats-reflect.md:5]: It's changed and not fixed because we had 2 cases that matched the new type and one that matched the old one.

**Comment by Mouvedia** [comment]
@ybiquitous I have split the PR as requested; can you take another look?

**Comment by Mouvedia** [comment]
@ybiquitous I found only one plugin rule in the wild that would be affected by this type change.
It's `@mrhenry/stylelint-mrhenry-nesting`](https://github.com/mrhenry/stylelint-mrhenry/blob/6939dd319ec125ae7ae9b3939bf106e9f39b452a/packages/nesting/index.mjs#L32) which is luckily being maintained by Romain; sorry @romainmenke :)
I am not surprised that the type itself being inconsistently used on our very own codebase wouldn't be adopted by plugin authors. It's obviously less useful in its current form and probably wasn't intended in the first place.

**Comment by Mouvedia** [comment]
@ybiquitous I found only one plugin rule in the wild that would be affected by this type change.
It's [`@mrhenry/stylelint-mrhenry-nesting`](https://github.com/mrhenry/stylelint-mrhenry/blob/6939dd319ec125ae7ae9b3939bf106e9f39b452a/packages/nesting/index.mjs#L32) which is luckily being maintained by Romain; sorry @romainmenke :)
I am not surprised that the type itself being inconsistently used on our very own codebase wouldn't be adopted by plugin authors. It's obviously less useful in its current form and probably wasn't intended in the first place.

**Comment by ybiquitous** [comment]
Thanks for the investigation. Luckily, this type change doesn't seem to have a big impact.

File: types/stylelint/index.d.ts [hunk 2]
```diff
@@ -246,7 +246,7 @@ declare namespace stylelint {
 	/** @internal */
 	export type RuleBase<P = any, S = any> = (
```
Comment (ybiquitous) [inline types/stylelint/index.d.ts:247]: [suggest] I'm not sure, but is it worth making the `S` default type more specific? For example:

```suggestion
	export type RuleBase<P = any, S = Record<string, any>> = (
```

File: .changeset/poor-boats-reflect.md (hunk 1 above)
Comment (ybiquitous) [inline .changeset/poor-boats-reflect.md:5]: [suggest] Minor. The following may be more explicit:

```suggestion
Changed: `secondaryOptions` argument type of the `Rule` function
```

Comment (ybiquitous) [reply .changeset/poor-boats-reflect.md:5]: FYI, the `Rule` type is exposed:
https://github.com/stylelint/stylelint/blob/18dff4f65360b20e434c572850a24d88e035c614/types/stylelint/index.d.ts#L271-L274

**Comment by romainmenke** [comment]
> It's [@mrhenry/stylelint-mrhenry-nesting](https://github.com/mrhenry/stylelint-mrhenry/blob/6939dd319ec125ae7ae9b3939bf106e9f39b452a/packages/nesting/index.mjs#L32) which is luckily being maintained by Romain; sorry @romainmenke :)

That is fine :)
I've removed these type hints for now:

https://github.com/mrhenry/stylelint-mrhenry/blob/884ac304d5a2fd5ed70591d12c5325a1c9e4eb3f/packages/nesting/index.mjs#L32

File: types/stylelint/index.d.ts (hunk 2 above)
Comment (Mouvedia) [reply types/stylelint/index.d.ts:247]: It's slightly more likely to be more error-prone but I agree with you nonetheless.
i.e. it's the de facto convention

Comment (Mouvedia) [reply types/stylelint/index.d.ts:247]: I don't think it should be the default.
Do you mean `secondaryOptions: S extends Record<string, any> ? S : never,` instead?

File: types/stylelint/index.d.ts [hunk 3]
```diff
@@ -246,7 +246,7 @@ declare namespace stylelint {
...
 		primaryOption: P,
-		secondaryOptions: Record<string, S>,
+		secondaryOptions: S,
```
Comment (Mouvedia) [inline types/stylelint/index.d.ts:249]:
```suggestion
		secondaryOptions: S extends Record<string, any> ? S : never,
```

see https://github.com/stylelint/stylelint/pull/7950/files#r1730844881

File: types/stylelint/index.d.ts (hunk 2 above)
Comment (ybiquitous) [reply types/stylelint/index.d.ts:247]: No. If my suggestion cannot get a value, let's leave it as-is. I was just curious.

File: types/stylelint/index.d.ts (hunk 3 above)
Comment (ybiquitous) [reply types/stylelint/index.d.ts]: Sorry for my late comment. Keeping it simple (i.e., `secondaryOptions: S`) seems enough for now.

File: types/stylelint/index.d.ts (hunk 2 above)
Comment (Mouvedia) [reply types/stylelint/index.d.ts:247]: Are you fine with [17cd6ca](https://github.com/stylelint/stylelint/pull/7950/commits/17cd6ca5eeb35b435539eeb9fd763f9eee600032) or should I revert to `S`.
It's slightly more strict.
I am not sure myself if we want to restrict the type for plugins or not.
i.e. maybe an array could be acceptable

File: .changeset/poor-boats-reflect.md (hunk 1 above)
Comment (Mouvedia) [reply .changeset/poor-boats-reflect.md:5]: Yes but RuleBase is internal :)

```ts
	/** @internal */
	export type RuleBase<P = any, S = any> = (
```

It flows back to `Rule`, you are right.
I thought I would be spared on a technicality.

File: types/stylelint/index.d.ts [hunk 4]
```diff
@@ -246,7 +246,7 @@ declare namespace stylelint {
 	/** @internal */
 	export type RuleBase<P = any, S = any> = (
 		primaryOption: P,
-		secondaryOptions: Record<string, S>,
+		secondaryOptions: S extends Record<string, any> ? S : never,
```
Comment (Mouvedia) [inline types/stylelint/index.d.ts:249]:
```suggestion
		secondaryOptions: S,
```

see https://github.com/stylelint/stylelint/pull/7950#discussion_r1730886345

**Review by ybiquitous — APPROVED** [review]
Thanks, LGTM 👍🏼
